doc/design/prune.md

```diff
@@ -0,0 +1,110 @@
+# Prune
+
+## Motivation
```

Comment:
where will Pruning happen, on the client or on the master?

Reply:
In the current design, prune is based only on `ProgramDesc`. So both designs are fine.